File: bookreviews/collect_together.py
import requests
from bs4 import BeautifulSoup

import os
import csv

BASEDOMAIN = "https://www.rtbookreviews.com"

def extract_links(pagenum):
    with open("pagedownloads/{}.html".format(pagenum)) as df:
        listpage = df.read()
    soup = BeautifulSoup(listpage, "html.parser")
    links = []
    for div in soup.find_all("div", class_="views-field-title"):
        links.append(div.span.a['href'])
    return links


def download_page(pagenum):
    r = requests.get("https://www.rtbookreviews.com/book-review?page={}".format(pagenum))
    with open("pagedownloads/{}.html".format(pagenum), 'w') as of:
        of.write(r.text)


def collect_reviews():
    for page in [123]:#range(1282):
        links = extract_links(page)
        for link in links:
            print(r)
            r = requests.get("https://www.rtbookreviews.com{}".format(link))
            #pagename=link.split(link, "/")[-1]
            pagename = link.replace("/book-review/", "")
            with open("reviewdownloads/{}.html".format(pagename), 'w') as of:
                of.write(r.text)


def process_reviews():
    reviewfiles = [os.path.join("reviewdownloads/", f)
                 for f in os.listdir("reviewdownloads/")
                 if os.path.isfile(os.path.join("reviewdownloads/", f)) and f.endswith(".html")]

    with open("results.csv", "w") as of:
        writer = csv.DictWriter(of, fieldnames=['url', 'title'])
        writer.writeheader()
        for rfile in reviewfiles:
            row = {}
            print(rfile)
            with open(rfile) as df:
                rhtml = df.read()
            soup = BeautifulSoup(rhtml, "html.parser")
            row['url'] = rfile
            row['title'] = soup.find("h1", class_="title").get_text()
            writer.writerow(row)




#download_page(123)
#print(extract_links(123))
#collect_reviews()

process_reviews()

File: bookreviews/collect_together.r
library(httr)
library(rvest)
library(glue)
library(stringr)

setwd("bookreviews")

BASEDOMAIN <- "https://www.rtbookreviews.com"

extract_links <- function(pagenum) {
  listpage <- read_html(glue("pagedownloads/{pagenum}.html"))
  listpage %>%
    html_nodes(".views-field-title span a") %>%
    html_attr("href") 
  
}

download_page <- function(pagenum) {
  r <- GET(glue("https://www.rtbookreviews.com/book-review?page={pagenum}"))
  write(content(r, as = "text"), file=glue("pagedownloads/{pagenum}.html"))
}

collect_reviews <- function() {
  for (page in c(342)) {#(page in 0:1281) {
    links <- extract_links(page)
    for (link in links) {
      print(link)
      r <- GET(glue("https://www.rtbookreviews.com{link}"))
      pagename <- str_replace(link, "/book-review/", "")
      write(content(r, as = "text"), file=glue("reviewdownloads/{pagename}.html"))
    }
  }
}

process_reviews <- function() {
  reviewfiles <- list.files(path="reviewdownloads", pattern="html$", full.names = TRUE)
  result <- data.frame(url=rep(NA, length(reviewfiles)), 
                       title=NA, author=NA)
  for (i in 1:length(reviewfiles)) {
    #print(reviewfiles[i])
    rhtml <- read_html(reviewfiles[i])
    result$url[i] <- reviewfiles[i]
    result$title[i] <- rhtml %>%
      html_node("h1.title") %>%
      html_text()
    result$author[i] <- rhtml %>%
      html_node("div.field-name-field-authors") %>%
      html_text()
    result$author[i] <- str_replace(result$author[i], "Author\\(s\\):\\s", "")
  }
  return(result)
}


#download_page(342)
#extract_links(342)
#collect_reviews()
process_reviews() %>% head()
